### ISSUE TITLE
RFE: add new X-mount.diskseq= synthetic mount option

### ISSUE THREAD

**Issue by poettering** [util-linux/util-linux#1786]
In trusted environments it often matters to mount a very specific incarnation of a block device, and minimize the chance that the block device is swapped out by an attacker in the time between it is found and mounted.

Recent kernels gained the "diskseq" concept for block devices. This contains a serial number that is is assigned to a block device whenever a new device pops up. i.e. if you plug in a usb stick that is exposed as /dev/sdb and then remove it it, and plug in another one that also will get /dev/sdb it will carry a different diskseq.

The diskseq is updated whenever a device is reused, as well as on media change. It is updated when a loopback device is attached, and when a new file is attached to it also updated.

I think it would be good if /bin/mount would support mounting a specific block device with a specific diskseq. Idea would be: a new mount option X-mount.diskseq= could be specified, taking a disk seq serial. if it is specified, /bin/mount would not pass the device node as is to the kernel mount() syscall, but open it first, then use the new `BLKGETDISKSEQ` ioctl to query the current diskseq of the device. if it matches the specified one it would then invoke the mount() syscall but replace the device path with a `/proc/self/fd/<fd>` path. If the number does not match it would immediately fail. Otherwise, once the mount() syscall finishes it would issue the `BLKGETDISKSEQ` command again, checking if the diskseq still matches. if so, it would exit successfully. If it doesn't anymore, the block device got swapped out in the short time window from before the mount() and after it. In that case it would fail, but first unmount the device again. (maybe make this atomic, by mounting in a private namespace first, and then moving it out only once complete).

(Alternative implementation: pass the diskseq to the kernel via the new mount API and make the kernel check the diskseq itself, but this would be politically more involved).

If this existed we'd use this in systemd's `gpt-auto` logic for example, which looks at the GPT table to make mounting decisions and where we should make sure that we actually mount the device we probed in the gpt part table and nothing else.

(Also see this discussion: https://github.com/systemd/systemd/issues/24503)

**Comment by karelzak** [comment]
I think X-mount.diskseq= together with a private namespace sounds good. 

Do you assume /dev/disk/by-diskseq/ ? If yes, we can trigger the functionality in libmount automatically if the mount source will be this symlink. So, users will not have to specify X-mount.diskseq= on mount(8) command line (or in fstab).

**Comment by poettering** [comment]
parsing the diskseq from the /dev/disk/by-diskseq/ path would make sense too I guess, but at least for my own usecase (i.e. the gpt-auto stuff) having only the explicit mount option would be sufficient.

(btw, i didn't test whether the populare block based file systems are actually OK with mounting a block device by fd via /proc/self/fd/…. I presume so, but I never tested it)

**Comment by DemiMarie** [comment]
> Do you assume /dev/disk/by-diskseq/ ? If yes, we can trigger the functionality in libmount automatically if the mount source will be this symlink. So, users will not have to specify X-mount.diskseq= on mount(8) command line (or in fstab).

That would be awesome indeed.  Could this check also be made for symlinks to `/dev/disk/by-diskseq/*`?

**Comment by karelzak** [comment]
I've tried to play with it. 

It works as expected, but the problem is the idea with the FD path, after the mount(2), you get "/proc/self/fd/3" in mountinfo, because the kernel never converts the source path to the real device path. It means the FD path will be everywhere in outputs of mount, findmnt, df, etc.

**Comment by DemiMarie** [comment]
What about using the new mount API?

**Comment by poettering** [comment]
i guess there might be file systems that don't like it if userspace has an fd open to a block device they are supposed to mount btw. the after-mount-recheck-of-the-diskseq should hence probably done via sysfs, not via ioctl.

i.e. `/sys/dev/block/<major>:<minor>/diskseq` will tell you the diskseq of the block device. but it sucks a bit since unlike the ioctl it doesn't actually work for partition block devices, only for the originating block devices, so you'd have to chase that through /sys/. but maybe you have code for that anyway?

**Comment by brauner** [comment]
I'll take a look what FSCONFIG_SET_FD currently does tomorrow.

**Comment by brauner** [comment]
I have a ToDo item for this to support a new "source-diskseq" property that can be specified via `fsconfig()`. That will involve filesystem work and conversion to the new mount api. But that shouldn't be a huge deal hopefully.

**Comment by karelzak** [comment]
The question is how filesystem drivers support FSCONFIG_SET_FD for "source". Now, this setting is done by string (usually path). 

I doubt there is any standardized way to do the same by the file descriptor. Note that "source" is traditionally whatever, and it's exclusively interpreted by the filesystem driver (see, for example, fuse or network filesystems).

**Comment by karelzak** [comment]
@brauner "source-diskseq" as a new fsconfig() property (independent of "source") sounds better than trying to do the same by "source" and file descriptor. We still need any string as the source to have something for mountinfo file.

**Comment by brauner** [comment]
Yeah, that's what I have in mind. I just want to add a new fsconfig() property independent of "source". However, "source-diskseq" would require "source" to be specified ofc. "source-diskseq" to make it clear that it's the diskseq of "source".
